Rubberduck.Parsing/Grammar/VBALexer.g4

```diff
@@ -298,8 +300,9 @@ SINGLEQUOTE : '\'';
 UNDERSCORE : '_';
 WS : [ \t];
 GUIDLITERAL : '{' [0-9A-F]+ '-' [0-9A-F]+ '-' [0-9A-F]+ '-' [0-9A-F]+ '-' [0-9A-F]+ '}';
-IDENTIFIER :  ~[[\]()\r\n\t.,'"|!@#$%^&*\-+:=; 0-9-/\\-] ~[[\]()\r\n\t.,'"|!@#$%^&*\-+:=; -]*;
+IDENTIFIER :  ~[[\](){}\r\n\t.,'"|!@#$%^&*\-+:=; 0-9-/\\-] ~[[\](){}\r\n\t.,'"|!@#$%^&*\-+:=; -]*;
 LINE_CONTINUATION : [ \t]* UNDERSCORE [ \t]* '\r'? '\n';
+BARE_HEX_LITERAL : [0-9a-fA-F]+;
```

Comment:
I really don't like this lexer rule. A lot of it's matching space is already covered by IDENTIFIER (which is terrible). It can't ever match hex literals starting with a letter, because those are already matched by the IDENTIFIER token.

Are we 100% sure we need this rule? And if yes, can we document what we need it for?

Reply:
This rule is necessary for the VB6 Form declarations. These use hex numbers without a format specifier after the `OleObjectBlob`, which is totally stupid in a language that has hex literals, btw. There is no other rule matching a number followed by a letter.

Unfortunately, legal hex numbers both have an overlap with decimal numbers and identifiers. None of the two other possibilities must be matchedby this rule. Otherwise this special case handling for VB forms would cause problems in ordinary VBA. (Remember that what has been matched in the lexer cannot be undone in the parser.)

Reply:
Could make it perhaps more targeted with `[0-9][0-9a-fA-F]*;`, that would at least eliminate the overlap and (perhaps) make the intent more explicit ? The name might need tweaking if so though.

Reply:
Am I missing something? Hex literals can start with A-F, not just 0-9, given a sufficient offset size into the FRX.

Reply:
Couldn't the leading colon be used as a way of parsing the offsetIdentifier?

    [\:]whitespace?[0-9a-fA-F]

Reply:
Since the lexer generally prefers longer matches, this would tokenize `bar : Add` as an identifier followed by an offset specification. 

Moreover, `whiteSpace` is not a thing in the the lexer. It is a parser rule.

BTW, we have the fragment `DIGIT` for `[0-9].

To explain why the suggestion to put the digit first: This removes the overlap with `IDENTIFIER`, which is currently preferred over `BARE_HEX_LITERAL`, as it should be, because it matches at least the same length and comes first in the lexer rules.
Offsets that do start with a letter are handled in the parser by also allowing `unrestrictedIdentifier`. (This could be simplified to `IDENTIFIER`, which is actually cleaner, but never came to 
my mind.) Hex numbers looking like decimal numbers are also not matched by `BARE_HEX_LITERAL` since they are already tokenized as `DECIMALLITERAL`. So we also allow `numberLiteral` for the offset. 

My suggestion would be to leave the rule as it is and add a comment what it is for and that numbers starting with a letter or only consisting of digits will be matched by other rules.

Reply:
@ThunderFrame just curious - have you ever seen an example with whitespace between the colon and the number? I haven't, and wonder if that would be legal.

If not, then the suggested lexer rule would become:

    RESOURCE_OFFSET : ':' [0-9a-zA-Z]+;

which makes the grammar:

    moduleConfigElement :
    (unrestrictedIdentifier | lExpression) whiteSpace? EQ whiteSpace? (shortcut | DOLLAR? expression RESOURCE_OFFSET?)? endOfStatement

This looks much cleaner to my eye. The only caveat is that `RESOURCE_OFFSET` would have to be defined before `COLON`, which I think should be fine.

Just testing this now - it has revealed a pre-existing issue with the Mid statement on this line:

https://github.com/rubberduck-vba/Rubberduck/blob/e2468806b7da88f20ca28a4ea0f08edb92cacd78/Rubberduck.Parsing/Grammar/VBAParser.g4#L467

IIUC, `ASSIGN` here is not correct, it should be `EQ`.

Reply:
This lexer rule is equally disrupting. The whitespace after a statement separator is completely optional. Try what happens to `bar:Add` with this rule present.

Reply:
Regarding the `Mid`, that should be an `EQ` indeed. Please fix that in another commit.